icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -821,14 +791,51 @@ MeasureUnitImpl MeasureUnitImpl::forMeasureUnitMaybeCopy(
 
 void MeasureUnitImpl::takeReciprocal(UErrorCode& /*status*/) {
     identifier.clear();
-    for (int32_t i = 0; i < units.length(); i++) {
-        units[i]->dimensionality *= -1;
+    for (int32_t i = 0; i < singleUnits.length(); i++) {
+        singleUnits[i]->dimensionality *= -1;
     }
 }
 
-bool MeasureUnitImpl::append(const SingleUnitImpl& singleUnit, UErrorCode& status) {
+bool MeasureUnitImpl::appendSingleUnit(const SingleUnitImpl &singleUnit, UErrorCode &status) {
     identifier.clear();
-    return appendImpl(*this, singleUnit, status);
+
+    if (singleUnit.isDimensionless()) {
+        // Do not append dimensionless units.
+        return false;
+    }
+
+    // Find a similar unit that already exists, to attempt to coalesce
+    SingleUnitImpl *oldUnit = nullptr;
+    for (int32_t i = 0; i < this->singleUnits.length(); i++) {
+        auto *candidate = this->singleUnits[i];
+        if (candidate->isCompatibleWith(singleUnit)) {
+            oldUnit = candidate;
+        }
+    }
+
+    if (oldUnit) {
+        // Both dimensionalities will be positive, or both will be negative, by
+        // virtue of isCompatibleWith().
+        oldUnit->dimensionality += singleUnit.dimensionality;
+
+        return false;
+    }
+
+    // TODO: question, shall we add a copy of singleUnit. I think we should, because the added singleUnit
```

Comment:
SingleUnit is cheap to copy (no memory allocation), so yes, we should copy it.

Reply:
done. Actually, it is copied already and we do not need the copy() function as in Java.